spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java

```diff
@@ -313,18 +310,50 @@ private void releaseMessageAfterDelay(final Message<?> message, long delay) {
 			this.messageStore.addMessageToGroup(this.messageGroupId, delayedMessage);
 		}
 
-		final Message<?> messageToSchedule = delayedMessage;
+		final UUID messageId = delayedMessage.getHeaders().getId();
 
 		this.getTaskScheduler().schedule(new Runnable() {
 
 			@Override
 			public void run() {
-				releaseMessage(messageToSchedule);
+				Message<?> messageToRelease = getMessageById(messageId);
+				if (messageToRelease != null) {
+					releaseMessage(messageToRelease);
+				}
 			}
 
 		}, new Date(messageWrapper.getRequestDate() + delay));
 	}
 
+	private Message<?> getMessageById(UUID messageId) {
+		Message<?> theMessage = null;
+		if (this.messageStore instanceof SimpleMessageStore) {
+			synchronized (this.messageGroupId) {
+				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
+				for (Message<?> message : messages) {
+					if (message.getHeaders().getId().equals(messageId)) {
+						theMessage = message;
+						break;
+					}
+				}
```

Comment:
This can be optimized to a simple `messages.get()` if the underlying `Collection` is a `Set`.